p2p/net/upgrader/upgrader.go

```diff
@@ -212,10 +210,14 @@ func (u *upgrader) setupMuxer(ctx context.Context, conn sec.SecureConn, server b
 
 	var smconn network.MuxedConn
 	var err error
-	// TODO: The muxer should take a context.
+	var streamMuxer *Multiplexer
+
 	go func() {
 		defer close(done)
-		smconn, err = u.muxer.NewConn(conn, server, scope)
+		streamMuxer, err = u.mstream.NegotiateMuxer(conn, server)
```

Comment:
I think `NegotiateMuxer` should take a ctx and then we don't have to do this select dance below and we don't have to spawn a go routine.